README.md
Validating:

- phone
- name
- postalCode
- email
- url
- currency
- ip
- date
- time
- htmlTags
- password (strong)
- password (medium)
- creditCard

For Dart/Flutter developers.
[Regexed Validator at pub.dartlang.org](https://pub.dartlang.org/packages/regexed_validator "Regexed Validator")

## Usage

Usage is pretty simple just pass a phone, date, email... to its validation method, see the example below

```dart
import 'package:regexed_validator/regexed_validator.dart';

main() {
  print( validator.email('example@example.co.uk') ); // true

  print(validator.url('http://www.nowhere.com?product=28&color=blue')); // true

  print(validator.currency('£498.10')); // true

  print(validator.ip('67.52.159.38')); // true

  print(validator.time('14:34 GMT -5')); // true

  print( validator.htmlTags('<strong>Bold</strong>\n'
          '<em>Emphazied</em>\n'
          '<b>Bold</b>\n'
          '<i>Italics</i>\n'
          '<span id="foo" class="bar">Some text</span>\n'
          '<hr />')
  ); // true
}
```

## With Flutter TextField, TextFormField and Form(s) 

There is nothing specific to this package when working with Forms and TextFields, see the example below
```dart
TextFormField(
    // The validator receives the text that the user has entered, and use regexed_validator to check if it is a valid string.
    validator: (value) {
      if (!validator.email(value)) {
        return 'Please enter a valid email';
      }
      return null;
    },
  ),
```

Here is a complete example, with a Form and an email field:
```dart
import 'package:flutter/material.dart';
import 'package:regexed_validator/regexed_validator.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    const appTitle = 'Form Validation Demo';

    return MaterialApp(
      title: appTitle,
      home: Scaffold(
        appBar: AppBar(
          title: const Text(appTitle),
        ),
        body: const MyCustomForm(),
      ),
    );
  }
}

// Create a Form widget.
class MyCustomForm extends StatefulWidget {
  const MyCustomForm({Key? key}) : super(key: key);

  @override
  MyCustomFormState createState() {
    return MyCustomFormState();
  }
}

// Create a corresponding State class.
// This class holds data related to the form.
class MyCustomFormState extends State<MyCustomForm> {
  // Create a global key that uniquely identifies the Form widget
  // and allows validation of the form.
  //
  // Note: This is a GlobalKey<FormState>,
  // not a GlobalKey<MyCustomFormState>.
  final _formKey = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    // Build a Form widget using the _formKey created above.
    return Form(
      key: _formKey,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          TextFormField(
            // The validator receives the text that the user has entered, and use regexed_validator to check if it is a valid string.
            validator: (value) {
              if (value == null || !validator.email(value)) {
                return 'Please enter a valid email';
              }
              return null;
            },
          ),
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 16.0),
            child: ElevatedButton(
              onPressed: () {
                // Validate returns true if the form is valid, or false otherwise.
                if (_formKey.currentState!.validate()) {
                  // If the form is valid, display a snackbar. In the real world,
                  // you'd often call a server or save the information in a database.
                  ScaffoldMessenger.of(context).showSnackBar(
                    const SnackBar(content: Text('Processing Data')),
                  );
                }
              },
              child: const Text('Submit'),
            ),
          ),
        ],
      ),
    );
  }
}
```


CHANGELOG.md
## 2.0.0+1
- Update README.md flutter example

## 2.0.0
- Increase the length for password (strong) validation regex check

## 1.0.4
- Increase the length for password (strong) validation regex check

## 1.0.3
- Add medium password strength, the strong one is still under bool password(String input)

## 1.0.2
- Update README.md usage example and lib description
- Update CHANGELOG.md
- Update pubspec.yaml description

## 1.0.1
- dartfmt

 ## 1.0.0

- Initial version, created with basic yet functional phone, name, postalCode, email, url, currency, ip, date, time, htmlTags, password, and creditCard validation.


lib/src/validator_base.dart
final RegExp _phoneRegex = RegExp(r"^[+]*[(]{0,1}[0-9]{1,4}[)]{0,1}[-\s\./0-9]*$");
final RegExp _nameRegex =
    RegExp(r"^([A-Z][A-Za-z.'\-]+) (?:([A-Z][A-Za-z.'\-]+) )?([A-Z][A-Za-z.'\-]+)$");
final RegExp _postalCodeRegex = RegExp(r"^\d{5}(-\d{4})?$");
final RegExp _emailRegex = RegExp(
    r'^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$');
final RegExp _urlRegex =
    RegExp(r"^(?:http|https):\/\/[\w\-_]+(?:\.[\w\-_]+)+[\w\-.,@?^=%&:/~\\+#]*$");
final RegExp _currencyRegex = RegExp(r"^(\$|\u00A3|\u00A5|\uFFE5)(\d*\.\d+|\d+)$");
final RegExp _ipRegex = RegExp(
    r"^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$");
final RegExp _dateRegex = RegExp(r"^\d{4}[-/](0?[1-9]|1[012])[-/](0?[1-9]|[12][0-9]|3[01])$");
final RegExp _timeRegex =
    RegExp(r"^([0-1]?[0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?( ([A-Z]{3}|GMT [-+]([0-9]|1[0-2])))?$");
final RegExp _htmlTagsRegex = RegExp(
    r"^<(?:([A-Za-z][A-Za-z0-9]*)\b[^>]*>(?:.*?)</\1>|[A-Za-z][A-Za-z0-9]*\b[^/>]*/>)$",
    multiLine: true);
final RegExp _passwordRegex =
    RegExp(r"^(?=.*\d)(?=.*[~!@#$%^&*()_\-+=|\\{}[\]:;<>?/])(?=.*[A-Z])(?=.*[a-z])\S{8,99}$");
final RegExp _mediumPasswordRegex = RegExp(
    r"^(((?=.*[a-z])(?=.*[A-Z]))|((?=.*[a-z])(?=.*[0-9]))|((?=.*[A-Z])(?=.*[0-9])))(?=.{6,})");
final RegExp _creditCardRegex = RegExp(
    r"^(?:3[47]\d{2}([\- ]?)\d{6}\1\d{5}|(?:4\d{3}|5[1-5]\d{2}|6011)([\- ]?)\d{4}\2\d{4}\2\d{4})$");

class _Validator {
  const _Validator();
  bool phone(String input) => _phoneRegex.hasMatch(input);
  bool name(String input) => _nameRegex.hasMatch(input);
  bool postalCode(String input) => _postalCodeRegex.hasMatch(input);
  bool email(String input) => _emailRegex.hasMatch(input);
  bool url(String input) => _urlRegex.hasMatch(input);
  bool currency(String input) => _currencyRegex.hasMatch(input);
  bool ip(String input) => _ipRegex.hasMatch(input);
  bool date(String input) => _dateRegex.hasMatch(input);
  bool time(String input) => _timeRegex.hasMatch(input);
  bool htmlTags(String input) => _htmlTagsRegex.hasMatch(input);
  bool password(String input) => _passwordRegex.hasMatch(input);
  bool mediumPassword(String input) => _mediumPasswordRegex.hasMatch(input);
  bool creditCard(String input) => _creditCardRegex.hasMatch(input);
}

final _Validator validator = const _Validator();


example/main.dart
import 'package:regexed_validator/regexed_validator.dart';

main() {
  print(validator.phone('+(123) - 456-78-90')); // true
  print(validator.name('John Quincy Adams')); // true
  print(validator.postalCode('10010-6543')); // ture
  print(validator.email('example@example.co.uk')); // true
  print(validator.url('http://www.nowhere.com?product=28&color=blue')); // true
  print(validator.currency('£498.10')); // true
  print(validator.ip('67.52.159.38')); // true
  print(validator.date('2000-12-20')); // true
  print(validator.time('14:34 GMT -5')); // true
  print(validator.htmlTags('<strong>Bold</strong>\n'
      '<em>Emphazied</em>\n'
      '<b>Bold</b>\n'
      '<i>Italics</i>\n'
      '<span id="foo" class="bar">Some text</span>\n'
      '<hr />')); // true
  print(validator.password('sword#42Fish')); // true
  print(validator.creditCard('4000-1234-1234-1234')); // true
}
